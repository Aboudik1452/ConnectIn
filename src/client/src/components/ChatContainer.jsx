import ContainerCSS from "../css/ChatContainer.module.css";
import { useState, useEffect, useRef } from "react";
import ChatInput from "./ChatInput";
import axios from "axios";

const ChatContainer = ({ currentChat, socket, room }) => {
  const [messages, setMessages] = useState([]);
  const [arrivalMessage, setArrivalMessage] = useState([]);
  const messagesEndRef = useRef(null);

  //Get id of logged in user
  const userID = sessionStorage.getItem("userID");

  useEffect(() => {
    fetchMessages();
  }, [currentChat]);

  useEffect(() => {
    messagesEndRef.current?.scrollIntoView({ behavior: "smooth" });
  }, [messages]);

  const fetchMessages = async () => {
    try {
      if (userID) {
        const response = await axios.post(
          `/messages`,
          {
            from: userID,
            to: currentChat.userID,
          }
        );
        const processedMessages = response.data.map((message) => {
          if (
            message.downloadLink &&
            message.downloadLink !==
              "https://connectin-api.onrender.com/messages/download/null"
          ) {
            message.message = (
              <a href={message.downloadLink} target="_blank" rel="noreferrer">
                {message.message }
              </a>
            );
          }
          return message;
        });
        setMessages(processedMessages);
      }
    } catch (error) {
      console.log(error);
    }
  };

  const handleSendMsg = async (msg, file) => {
    const data = new FormData();
    data.append("message", msg.trim());
    data.append("from", userID);
    data.append("to", currentChat.userID);
    data.append("room", room);
    data.append("value", new Date(Date.now()));
    data.append("file", file);
  
    if (msg.trim().length !== 0 || file) {
      // emit message to server using socket connection
      socket.current.emit("sendMessage", {
        message: msg.trim(),
        from: userID,
        to: currentChat.userID,
        room: room,
        value: new Date(Date.now()),
        file: file,
      });
  
      try {
        const response = await axios.post(
          "/messages/addMessage",
          data
        );
  
        let messageObject = {
          fromSelf: true,
          message: msg.trim(),
          file: response.data,
        };
  
        if (response.data.downloadLink) {
          messageObject = {
            fromSelf: true,
            message: (
              <a
                href={response.data.downloadLink}
                target="_blank"
                rel="noreferrer"
              >
                {response.data.filename}
              </a>
            ),
            file: response.data,
          };
        }
  
        setMessages(prevMessages => [...prevMessages, messageObject]);
      } catch (error) {
        console.log(error);
      }
    }
  };

  useEffect(() => {
    socket.current.on("receiveMessage", (message) => {
      setArrivalMessage({
        fromSelf: false,
        message: message.message,
        file: message.file,
      });
    });
  }, [socket]);

  useEffect(() => {
    if (arrivalMessage) {
      setMessages((prevMessages) => [...prevMessages, arrivalMessage]);
      setArrivalMessage(null);
    }
  }, [arrivalMessage]);
  
  return (
    <div className={ContainerCSS.container}>
      <div className={ContainerCSS.header}>
        <span>{currentChat.firstname} {currentChat.lastname}</span>
      </div>
      <hr className={ContainerCSS.line} />
      <div className={ContainerCSS.chatBox}>
        
      {messages.map((message, index) => {
        const isLastMessage = index === messages.length - 1;
        return (
          <div className={ContainerCSS.messagesContainer} key={index}>
            {message.fromSelf ? (
              <div className={ContainerCSS.sent}>
                <div className={ContainerCSS.textSent}>{message.message}</div>
              </div>
            ) : (
              <div className={ContainerCSS.received}>
                <div className={ContainerCSS.textReceived}>{message.message}</div>
              </div>
            )}
            {isLastMessage && <div ref={messagesEndRef} />}
          </div>
        );
      })}

      </div>
      <hr className={ContainerCSS.line} />
      <div className={ContainerCSS.input}>
        <ChatInput handleSendMsg={handleSendMsg} from={userID} to={currentChat.userID} />
      </div>

    </div>
    
  ); 
}
export default ChatContainer;