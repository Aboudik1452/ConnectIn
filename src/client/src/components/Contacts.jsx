import React, {useState} from "react";
import ContactCSS from "../css/Contacts.module.css";
import Avatar from '@mui/material/Avatar'

const Contacts = ({connections, changeChat}) => {
  const [currentSelected, setCurrentSelected] = useState(undefined);

    const changeCurrentChat = (index, contact) => {
      console.log(index);
      console.log(contact);
      setCurrentSelected(index)
      changeChat(contact)
    }
    const [loadChat, setLoadChat] = useState(false);
    
    const loadChatComponent = () => {
      setLoadChat(true);
    }

    return(
        <div className={ContactCSS.contacts}>
            <h2 className="mycontact">My Contacts</h2>
            <div className="ul">
             {connections && (connections.map((contact, index) => {
               return (
                
                 <div key={contact._id} className={ContactCSS.connectionInfo} onClick={() => changeCurrentChat(index, contact)}>
                   {/* <img className={ContactCSS.contactImage}
                     src="https://i.pinimg.com/originals/f1/0f/f7/f10ff70a7155e5ab666bcdd1b45b726d.jpg"
                     alt="comapnyPic"
                   /> */}
                   <Avatar alt="Logo" src="./logo/logo.png" size="medium"sx={{ width: 75, height: 75 }}/>

                   

                    <span className={ContactCSS.connectionName}>{contact.firstname} {contact.lastname}</span>

                 </div>
               );
             }))}
           </div>
        </div>
    )
}

export default Contacts;