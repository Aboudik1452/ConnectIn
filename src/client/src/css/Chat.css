/* .conatainer {
  display: flex;
  flex-direction: row;
}

.chatbox {
  height: 85vh;
  width: 68vw;
  background-color: #e0e0e099;
  margin-bottom: 80px;
  margin-top: 20px;
  margin-left: 20px;
  border-radius: 7%;
  display: flex;
  justify-content: flex-start;
  @media screen {
    min-width: 720px;
    max-width: 1080px;
  }
}

.contacts {
  height: 85vh;
  width: 45vh;
  background-color: #07bae054;
  right: 25px;
  position: absolute;
  top: 100px;
  margin-bottom: 80px;
  border-radius: 7%;
}

.userpic {
  width: 160px;
  height: 160px;
  border-radius: 80px;
  background-color: pink;
  margin: inherit;
}

.username {
  display: flex;
  justify-content: center;
  margin: auto;
  margin-top: 175px;
  font-weight: 600;
  font-style: normal;
}

.vl {
  border-left: 2px solid #00000012;
  height: 625px;
  position: absolute;
  left: 220px;
  margin-bottom: 30px;
  margin-top: 30px;
}

.mycontact {
  display: flex;
  justify-content: center;
  margin: 60px;
  font-weight: 600;
  font-style: normal;
}

.connectionsInfo {
  display: flex;
  border-radius: 30px;
  background-color: aliceblue;
  width: 300px;
  margin-top: 20px;
  margin-bottom: 20px;
  justify-content: flex-start;
  align-items: center;
}

.connectionName {
  margin-left: 40px;
  font-size: larger;
}

.ul {
  display: flex;
  flex-direction: column;
  padding-inline-start: 40px;
}

.message-container {
  max-width: max-content;
  display: flex;
  align-items: flex-end;
  left: 15px;
  position: relative;
}

.message {
  margin-bottom: 30px;
  height: 45px;
  width: 640px;
  border: none;
  border-radius: 20px;
  margin-left: 35px;
  background: #0000001f;
  outline: none;
  padding-left: 10px;
}

.sendicon {
  margin-bottom: 45px;
  color: rgba(0, 0, 0, 0.651);
  margin-left: 10px;
}

.attachicon {
  margin-bottom: 45px;
  color: rgba(0, 0, 0, 0.651);
  margin-left: 6px;
}

.sender-userpic {
  height: 40px;
  width: 40px;
  background-color: pink;
  border-radius: 20px;
  top: 454px;
  left: 35px;
  position: absolute;
}

.sender-text {
  height: max-content;
  width: max-content;
  background-color: #0000001f;
  border-radius: 20px;
  position: absolute;
  top: 454px;
  left: 35px;
  padding: 10px;
  margin-left: 50px;
  display: inline-block;
}

.sender-message {
  display: flex;
  left: 10px;
  position: relative;
}

.user-userpic {
  height: 40px;
  width: 40px;
  background-color: lightslategray;
  border-radius: 20px;
  top: 540px;
  left: 394px;
  position: absolute;
}

.user-text {
  height: max-content;
  width: max-content;
  background-color: #0000001f;
  border-radius: 20px;
  position: absolute;
  top: 540px;
  left: 88px;
  padding: 10px;
  display: inline-block;
}

.user-message {
  display: flex;
  left: 330px;
  position: relative;
} */
