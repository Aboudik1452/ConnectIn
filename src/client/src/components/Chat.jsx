import React, { useState, useContext, useEffect, useRef } from "react";
import "../css/Chat.css";
import { Context } from "../UserSession";
import axios from "axios";
import io from "socket.io-client";
import ChatCSS from "../css/Chat.module.css";
import ContactCSS from "../css/Contacts.module.css";
import { Container } from "@mui/system";
import Contacts from "./Contacts";
import ChatContainer from "./ChatContainer";

const socket = io.connect("http://localhost:9000")

const Chat = () => {
  
  const socket = io.connect("http://localhost:9000")
  

  // const sendMessage = () => {
  //   socket.emit()
  // };

  const [userConnections, setUserConnections] = useState(
    {
      _id: "",
      firstname: "",
      lastname: "",
      connections: [],
    }
  );
   //Global loginState
   const [login, setLogin] = useContext(Context);
   const [message, setMessage] = useState("");
  //  const [room, setRoom] = useState("");
   const scrollRef = useRef();
   const from = "6402817b781fecbdadf6c992";
   const to = "6402a4c6db0c6f36e8f531a8";
   const room = "room1"

  
   const [currentChat, setCurrentChat] = useState(undefined);

   //Get id of logged in user
   const userID = sessionStorage.getItem("userID");
 
   useEffect(() => {
     if (userID) {
       fetchSession();
       fetchUserConnections();
     }
   }, []);

      //Having the loginState persist on all page
      const fetchSession = async () => {
        try {
          if (userID) {
            setLogin({
              isLoggedIn: true,
            });
          }
        } catch (error) {
          console.log(error);
        }
      };
   
   const fetchUserConnections = async () => {

    try{
      if(userID){
        const response = await axios.get(
          `http://localhost:9000/users/profile/${userID}`
        );

        setUserConnections({
          _id: response.data._id,
          firstname: response.data.firstname,
          lastname: response.data.lastname,
          connections: response.data.connections,

        });
        // console.log(response.data);
      }
    }catch (error) {
      console.log(error);
   }
  };

 const sendMessage = ()=>{

  const data = {
    message: message,
    from: from,
    to: to,
    room: room,
    value: new Date(Date.now())
  }
  if(message.length !=0){
    socket.emit("sendMessage", data);
  }
    joinRoom();
    setMessage("");
 }

 const joinRoom = () =>{

  socket.emit("joinRoom", room);
 }

 useEffect(()=>{
    socket.on("receiveMessage", (data)=>{
      console.log(data.message)
    })
 }, [socket])

 const handleChangeChat = async (chat) => {
  setCurrentChat(chat)
  console.log(chat.userID)
  console.log(userID);
 await axios.post("http://localhost:9000/rooms",{
    userID_1: sessionStorage.getItem("userID"),
    userID_2: chat.userID
  }).then((response) =>{
    socket.emit("joinRoom", response.data)
  })
}
 

//  const joinRoom = () =>{
//   socket.emit("join_room", room);
//  }


  return (
    <>
    <Container>
      <div className={ChatCSS.container}>
        {/* {console.log(userConnections.connections)} */}
        <div>
          <Contacts connections={userConnections.connections} changeChat={handleChangeChat} />
        </div>
        <hr className={ChatCSS.line} /> 
        {/* <div className={ChatCSS.chat_container}> */}
        {currentChat === undefined ? (
          <div className={ContactCSS.selectHeader}>Select a chat</div>
        ) : (
          <ChatContainer currentChat={currentChat} socket={socket}/>
        )}
        </div>

      {/* </div> */}
    </Container>
    </>
  );
};

export default Chat;