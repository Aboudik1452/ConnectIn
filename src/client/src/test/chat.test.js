import React from "react";
import renderer from 'react-test-renderer'
import { screen, render } from "@testing-library/react";
import mockio, { io } from "socket.io-client";
import UserSession from "../UserSession";
import '@testing-library/jest-dom';
import Chat from '../components/Chat'
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import Contacts from "../components/Contacts";
import ChatContainer from "../components/ChatContainer";
import SocketMock from 'socket.io-mock';


beforeAll(() => {
    // mock.reset();
    sessionStorage.setItem("userID", '63ffcb1818750c7838b46a7c');
    sessionStorage.setItem("firstname", 'admin');
    sessionStorage.setItem("lastname", 'admin');
    sessionStorage.setItem("role", 'Administrator');
})

let mock = new MockAdapter(axios)

describe('validate whether or not the front-end of the chat page renders correctly', () => {

    const uID = sessionStorage.getItem('userID')
    mock.onGet(`/users/profile/${uID}`).reply(201, {
        connections: [{
            userID: '63ffca4118750c7838b46a68',
            firstname: 'Test',
            lastname: 'lastName',
        }]
    })

    const connection = [
        {
            "userID": '63ffca4118750c7838b46a68',
            "firstname": 'Test',
            "lastname": 'lastName',
        },
        {
            "userID": '63ffca4118750c7838b46a62',
            "firstname": 'John',
            "lastname": 'Doe',
        }
    ]

    it('renders static elements correctly', () => {
        const tree = renderer.create(
        <UserSession>
            <MemoryRouter>
                <Chat />
            </MemoryRouter>
        </UserSession>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    })

    it('renders dynamic components correctly', () => {
        const chatPage = render(
            <UserSession>
                <MemoryRouter>
                    <Contacts connections={connection}/>
                </MemoryRouter>
            </UserSession>
        );

        const contact1 = screen.getByText(/Test/i);

        expect(contact1).toBeInTheDocument();
        // const contacts = screen.getByTestId('contacts-container');
        // expect(contacts).toBeInTheDocument();
    })
    it('renders chatContainer correctly', () => {
        const chat = {
            "userID": '63ffca4118750c7838b46a68',
            "firstname": 'Test',
            "lastname": 'lastName',
        }
        const socket = new SocketMock();

        socket.on('sendMessage', (data) => {
            expect(data).toEqual('Hello World');
        });

        socket.on("receiveMessage", (data) => {
            expect(data).toEqual('Hello World');
        })

        socket.socketClient.emit('sendMessage', 'Hello World');
        socket.socketClient.emit('receiveMessage', 'Hello World');
        console.log(socket);

        const chatPage = render(
            <UserSession>
                <MemoryRouter>
                    <ChatContainer socket={socket} currentChat={chat}/>
                </MemoryRouter>
            </UserSession>
        );

        // const contacts = screen.getByTestId('contacts-container');
        // expect(contacts).toBeInTheDocument();
    })

    

})

afterAll(() => {
    sessionStorage.removeItem('userID');
    sessionStorage.removeItem('firstName');
    sessionStorage.removeItem('lastName');
    sessionStorage.removeItem('role');
    mock.reset();
});  