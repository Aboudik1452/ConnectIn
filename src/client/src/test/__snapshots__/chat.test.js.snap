// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate whether or not the front-end of the chat page renders correctly renders static elements correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="container"
  >
    <div
      data-testid="contacts-container"
    >
      <div
        className="contacts"
      >
        <div
          className="header"
        >
          My Contacts
        </div>
        <hr
          className="line"
        />
        <div
          className="ul"
        />
      </div>
    </div>
    <hr
      className="line"
    />
    <div
      className="selectHeader"
    >
      Select a chat
    </div>
  </div>
</div>
`;
